compiler/src/iree/compiler/Codegen/Common/GPU/GPUPackToIntrinsics.cpp

```diff
@@ -121,6 +121,119 @@ struct ConvertToMultiMma final : OpInterfaceRewritePattern<linalg::LinalgOp> {
   }
 };
 
+// This pattern hoists pack & unpack ops out of scf.for op.
+struct PackDestinationForOp final : OpRewritePattern<scf::YieldOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(scf::YieldOp yieldOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = yieldOp.getLoc();
+
+    // Get the enclosing scf.for op.
+    auto parentOp = yieldOp->getParentOp();
+    auto forOp = dyn_cast<scf::ForOp>(parentOp);
+    if (!forOp)
+      return failure();
+
+    linalg::UnPackOp unpackOp;
+    linalg::PackOp packOp;
+    int64_t tiedResultIdx = 0;
+
+    // Iterate through all operands of yieldOp & hoist each available
+    // pack-unpack pair.
+    for (auto operand : yieldOp.getOperands()) {
+      unpackOp = operand.getDefiningOp<linalg::UnPackOp>();
+      if (!unpackOp) {
+        tiedResultIdx++;
+        continue;
+      }
+
+      // Apply the pattern only if packOp & unpackOp are the only 2 users of the
+      // regionIterArg.
+      auto iterArg = forOp.getRegionIterArgs()[tiedResultIdx];
+      if (iterArg.getNumUses() != 2) {
+        tiedResultIdx++;
+        continue;
+      }
+
+      // Get the corresponding packOp.
+      for (auto user : iterArg.getUsers()) {
+        packOp = dyn_cast<linalg::PackOp>(user);
+        if (packOp &&
+            ((packOp.getInnerDimsPos() == unpackOp.getInnerDimsPos()) &&
+             (packOp.getMixedTiles() == unpackOp.getMixedTiles()) &&
+             (packOp.getOuterDimsPerm() == unpackOp.getOuterDimsPerm()))) {
+          break;
+        } else {
+          packOp = nullptr;
+        }
+      }
+      if (packOp && unpackOp) {
+        break;
+      }
+      tiedResultIdx++;
+    }
+    if (!packOp || !unpackOp) {
+      return failure();
+    }
+
+    // Create the pack -> new scf.for -> unpack chain.
+    rewriter.setInsertionPoint(forOp);
+    Value input = linalg::PackOp::createDestinationTensor(
+        rewriter, loc, forOp.getInitArgs()[tiedResultIdx],
+        packOp.getMixedTiles(), packOp.getInnerDimsPos(),
+        packOp.getOuterDimsPerm());
+
+    auto packedDest = rewriter.create<linalg::PackOp>(
+        loc, forOp.getInitArgs()[tiedResultIdx], input,
+        packOp.getInnerDimsPos(), packOp.getMixedTiles(),
+        packOp.getPaddingValue(), packOp.getOuterDimsPerm());
+
+    auto packOpValues = llvm::to_vector_of<Value>(forOp.getInitArgs());
+    packOpValues[tiedResultIdx] = packedDest.getResult();
+    scf::ForOp newForOp = rewriter.create<scf::ForOp>(
```

Comment:
If you are going to replace the old forop with the new forop, can you still iterate on the old for op to do further replacements? This is why I was thinking of doing this in two phases, in first phase you collect all pack/unpack pairs and idx to a struct, and then you replace all of them in a new forop in a single step.